lumicks/pylake/tests/test_channels/test_channels.py

```diff
@@ -116,23 +116,63 @@ def test_calibration_continuous_channels():
 
 
 def test_slice_properties():
-    size = 5
-    s = channel.Slice(channel.TimeSeries(np.random.rand(size), np.random.rand(size)))
+    size = 10
+    rng = np.random.default_rng()
+    timestamps_var = np.unique(np.sort(rng.integers(1e9, 10e9, size=size)))
+    timestamps_cst = np.arange(0, size)
+
+    # `Continuous`
+    s = channel.Slice(channel.Continuous(rng.random(size), start=0, dt=1))
     assert len(s) == size
-    assert s.sample_rate is None
+    assert s.sample_rate == 1e9
+    timesteps = np.asarray([1], dtype=int)
+    np.testing.assert_equal(s._timesteps, timesteps)
 
-    s = channel.Slice(channel.Continuous(np.random.rand(size), start=0, dt=1))
+    # `TimeSeries` with constant sample rate
+    s = channel.Slice(channel.TimeSeries(rng.random(size), timestamps_cst))
     assert len(s) == size
     assert s.sample_rate == 1e9
+    timesteps = np.unique(np.diff(timestamps_cst)).astype(int)
+    np.testing.assert_equal(s._timesteps, timesteps)
 
-    size = 10
+    # `TimeSeries` with variable sample rate
+    s = channel.Slice(channel.TimeSeries(rng.random(len(timestamps_var)), timestamps_var))
+    assert len(s) == len(timestamps_var)
+    with pytest.raises(
+        ValueError,
+        match="The sample rate is variable",
+    ):
+        s.sample_rate
+    timesteps = np.unique(np.diff(timestamps_var)).astype(int)
+    np.testing.assert_equal(s._timesteps, timesteps)
+
+    # `TimeTags`
     s = channel.Slice(channel.TimeTags(np.arange(0, size, dtype=np.int64)))
     assert len(s) == size
-    assert s.sample_rate is None
+    with pytest.raises(
+        NotImplementedError,
+        match="`sample_rate` is not available for TimeTags data",
+    ):
```

Comment:
Indeed handling of the `AttributError` is kinda cumbersome. It also is not optimal introducing another breaking change. I therefore reverted the functionality back to how it was and return a `None` in case no `sample_rate` is defined.

However, for `TimeSeries` data it still raises an `AttributeError` if the `sample_rate` is variable. I think otherwise users might be confused, when their code accessing `sample_rate` was usually working and then suddenly they are confronted with an empty `None` sample rate without an explanation. In this case maybe changing the `AttributeError` into a warning would make sense?

Reply:
I think this could also be explained in the docstring though. When they get `None`, the first thing they will (/should) probably look at is the docstring.

While I'm not entirely opposed to a warning, this would get annoying if you wrote your code to actually gracefully deal with this case. Since you'd still emit the warning even if all they're doing is checking if it's `None`. Sure you can then filter that warning, but that's also not so nice.

Reply:
OK, convinced 😄